Explain Kubernetes RBAC more comprehensively

ptgott opened the pull request:
Closes #18230

This change explains Teleport's RBAC system in relation to the Teleport Kubernetes Service. The docs didn't leave me with a readily available place to put this information, since explanations of Kubernetes-related role fields are scattered in asides throughout various pages.

As a result, I have added a more comprehensive, focused explanation of Teleport's Kubernetes RBAC system in two places:

- A new how-to guide for setting up Kubernetes RBAC with Teleport using a local `minikube` cluster.
- A reference guide that explains the Kubernetes-related fields within a Teleport role as well as the logic that the Kubernetes Service uses to evaluate these fields. This replaces the existing Kubernetes Access Controls guide.

To make room for the RBAC how-to guide, I've reorganized the other guides into a logical sequence: enabling auto-discovery, manually registering clusters, and setting up access controls.

I've also edited the Resource Access Requests guides to explain role fields for the different resources you can restrict access to via `search_as_roles`, including `pods`.

ptgott committed:
Explain Kubernetes RBAC more comprehensively

Closes #18230

This change explains Teleport's RBAC system in relation to the Teleport
Kubernetes Service. The docs didn't leave me with a readily available
place to put this information, since explanations of Kubernetes-related
role fields are scattered in asides throughout various pages.

As a result, I have added a more comprehensive, focused explanation of
Teleport's Kubernetes RBAC system in two places:

- A new how-to guide for setting up Kubernetes RBAC with Teleport using
  a local `minikube` cluster.
- A reference guide that explains the Kubernetes-related fields within a
  Teleport role as well as the logic that the Kubernetes Service uses to
  evaluate these fields. This replaces the existing Kubernetes Access
  Controls guide.

To make room for the RBAC how-to guide, I've reorganized the other
guides into a logical sequence: enabling auto-discovery, manually
registering clusters, and setting up access controls.

I've also edited the Resource Access Requests guides to explain role
fields for the different resources you can restrict access to via
`search_as_roles`, including `pods`.

tigrato commented on docs/pages/access-controls/access-requests/resource-requests.mdx:
````diff
@@ -323,52 +340,169 @@ spec:
+  deny: {}
+```
+
+##### Searching for pod resources
````
Searching will be introduced soon but currently, we don't support it.

ptgott replied:
I've renamed this section so we don't imply we support `tsh request search --kind=pods`

tigrato commented on docs/pages/access-controls/access-requests/resource-requests.mdx:
````diff
@@ -323,52 +340,169 @@ spec:
+/TELEPORT_CLUSTER/pod/KUBE_CLUSTER/NAMESPACE/POD_NAME
+```
+
+For example, to request access to a pod called `nginx-1` in the `development`
````
Mention that both `NAMESPACE` and  `POD_NAME` support regex and `*`

tigrato commented on docs/pages/includes/role-spec.mdx:
```diff
@@ -141,6 +141,19 @@ spec:
+      - kind: pod
+        # The name of the Kubernetes namespace in which to allow access the
+        # resources you specify with "name" and "kind". 
+        # The wildcard character "*" matches any sequence of characters.
```
we also support regex logic using the format `^REGEX_EXPRESSION$`

tigrato commented on docs/pages/reference/cli.mdx:
```diff
@@ -2406,7 +2406,8 @@ tctl version
 ## Resource filtering
 
-Both `tsh` and `tctl` allow you to filter `nodes`, `apps`, `db`, and `kube` resources using the `--search` and `--query` flags.
+Both `tsh` and `tctl` allow you to filter `nodes`, `apps`, `db`, `pod`, and
```
`tctl` does not allow you to search by `pod` since Pod resources do not exist in Teleport cache.
`tsh` does not support it yet but it will soon.

tigrato commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
```diff
@@ -0,0 +1,427 @@
+  - name: nginx
+    image: nginx:1.23
+---
+apiVersion: v1
```
In this example one can be a deployment so you can show the full power of Pod RBAC using wildcards since Kubernetes does not allow that in RBAC

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -1,236 +1,120 @@
-    kubernetes_users: ["{{external.email}}"]
+    kubernetes_labels:
+      'region': '*'
+      'environment': 'development'
```
maybe it's worth mentioning that a label key supports multiple values
`'key': ['val1','val2']`
You are allowed to access that cluster if the Kubernetes cluster has a label with `key=key` whose value is `val1` or `val2`

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -242,29 +126,81 @@ Label each cluster with key-value pairs describing the cluster:
-Here is an example of Kubernetes `ClusterRole` and `ClusterRoleBinding`
-allowing Teleport's `ServiceAccount` to impersonate users and groups:
+If a user has exactly one value in `kubernetes_users`, the Teleport Kubernetes
+Service impersonates that user. Otherwise, the Kubernetes Service denies the
```
If the user has no `kubernetes_users` his username is used as the user to impersonate.

Teleport denies the request if:
- user has multiple kubernetes_users assigned. He must select one of them with `--as`
- user tries to impersonate a group/user he is not allowed to impersonate


The last sentence here is not clear

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -242,29 +126,81 @@ Label each cluster with key-value pairs describing the cluster:
 
-Teleport uses [Kubernetes impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation)
-to map OIDC and SAML users to Kubernetes users and groups.
+If the user has not explicitly determined a Kubernetes user and Kubernetes
```
Mention that:
- Only one kubernetes_user is allowed to be impersonated per request. If the role allows multiple users, you must select one using `--as` otherwise the request is denied.
- When the role/user defines no `kubernetes_users`, Teleport will use the Teleport username as the user to impersonate.

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -242,29 +126,81 @@ Label each cluster with key-value pairs describing the cluster:
+
+#### Enabling impersonation
+
+To enable the Kubernetes Service to forward user requests with impersonation
```
Mention the script that creates the SA automatically with the roles, so users don't try to recreate this

ptgott replied:
I moved the `Notice` that was below the snippet above it so users see it before the snippet.

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+    # For Alice, will be substituted with the list ["admins", "devs"]
+    kubernetes_groups: ["{{external.groups}}"]
+    # For Alice, will be substituted with ["alice@example.com"]
+    kubernetes_users: ["{{external.email}}"]
```
Add an example for user traits

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+metadata:
+  name: kube-access
+version: v6
+spec:
```
you must also specify the `kubernetes_labels`. The access is only granted if the cluster satisfies the `kubernetes_labels` and the pod satisfies the `kubernetes_resources`

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+- `name`: The name of the pod to allow access to. In `kube-access`, this is
+  the `webapp` pod. You can add a wildcard character (`*`) to replace any
+  sequences of characters. For example, `name: "pod-*-*"` matches pods named
+  `pod-1-a` and `pod-2-c`.
```
Mention that we also support regex

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+### Authorizing user requests
+
+When the Teleport Kubernetes Service receives a request, it evaluates two fields
+within the user's roles. If these fields do not allow the user to perform the
```
```suggestion
within the user's roles. If one of these fields do not allow the user to perform the
```

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+do not match the user's request, the Kubernetes Service checks the next role,
+and so on. 
+
+If the Kubernetes Service finds a role with a `spec.allow` condition that
```
Maybe it's worth mentioning that by using `kubernetes_labels` and `kubernetes_resources` you can leverage different access levels for pods by using different `kubernetes_groups` and `kubernetes_users`. We collect the allowed `kubernetes_users` and `kubernetes_groups` using the same logic we use for allowing access to a pod/cluster.

tigrato commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -303,20 +239,295 @@ roleRef:
+<Notice type="warning">
 
-Integrate with your identity provider:
+If you have namespace restrictions in your Teleport RBAC, you should
```
Add the example of the fixed role so it's clear

```yaml
kind: role
metadata:
  name: kube-access-1
version: v6
spec:
  allow:
    kubernetes_groups:
      - default-pod-viewer
    kubernetes_resources:
      - kind: pod
        namespace: "default"
        name: "*"
    # ...
```

tigrato reviewed (commented):
I really like this total refactor from the Kubernetes docs.

I think it got much more accessible and simple!
Nice work @ptgott !

ptgott committed:
Partially respond to PR feedback

- Add `kubernetes_labels` to example snippets
- Rename a misleading H5 in the Resource Requests guide
- Make the description of impersonation more accurate in the Access
  Controls guide
- Move a Notice in the Access Controls guide
- Add an example role to the warning re: namespace restrictions in the
  Access Controls guide
- Make resource filtering info more accurate in the CLI reference

ptgott commented:
## To do

- [ ] Describe RegExp support in `kubernetes_resources` (https://github.com/gravitational/teleport/pull/20730#discussion_r1089207721, https://github.com/gravitational/teleport/pull/20730#discussion_r1089211756, https://github.com/gravitational/teleport/pull/20730#discussion_r1089244342)
- [ ] Use a `Deployment` to showcase wildcards in the how-to guide (https://github.com/gravitational/teleport/pull/20730#discussion_r1089223368)
- [ ] Describe support for multiple label values (https://github.com/gravitational/teleport/pull/20730#discussion_r1089232224)
- [ ] Add an example for user traits (https://github.com/gravitational/teleport/pull/20730#discussion_r1089241385)
- [ ] Describe allowing different pod access levels using Teleport RBAC (https://github.com/gravitational/teleport/pull/20730#discussion_r1089258191)

ptgott committed:
Change the RBAC how-to guide

Use Deployments in the demo cluster instead of Pods, showcasing the use
of regular expressions in per-pod RBAC.

This responds to PR feedback.

ptgott committed:
Provide an example of internal traits

This responds to PR feedback

ptgott committed:
Add notes about regex support

Responds to PR feedback

ptgott committed:
Explain multiple values in labels

Responds to PR feedback

zmb3 commented on docs/pages/kubernetes-access/controls.mdx:
```diff
@@ -18,6 +18,44 @@ guide](./manage-access/rbac.mdx).
+      'region': '*'
+      'platform': 'minikube'
+    kubernetes_resources:
+      - kind: pod
```
This is only applicable to v12+, right?

tigrato replied:
Yes

tigrato reviewed (approved):
I am quite happy with this guide!
Nice work @ptgott 

Regarding the different access levels per pod, I would recommend something like this:
```yaml
kind: role
metadata:
  name: kube-viewer
version: v6
spec:
  allow:
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: pod
        namespace: "*"
        name: "*"
    kubernetes_groups:
    - viewer # can list pods, and get them but exec and logs are forbidden
---
kind: role
metadata:
  name: kube-exec
version: v6
spec:
  allow:
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: pod
        namespace: "*"
        name: "nginx-*"
    kubernetes_groups:
    - execAndLogs
```

This would allow any user with both roles to list all pods in the cluster but can only exec and retrieve the logs for pods prefixed with `nginx-*`

ptgott commented:
@tigrato Thanks! I'll work that into the docs. 

@alexfornuto Let me know if you're planning to give this a copy review!

alexfornuto commented:
> @alexfornuto Let me know if you're planning to give this a copy review!

If you want one, well buddy you got it!

@tigrato are all the unresolved comments from your review now resolved?

tigrato commented:
> > @alexfornuto Let me know if you're planning to give this a copy review!
> 
> If you want one, well buddy you got it!
> 
> @tigrato are all the unresolved comments from your review now resolved?

@alexfornuto yes, except the last topic that @ptgott is still working on

ptgott committed:
Document progressively enabling access to pods

In response to PR feedback

Commit:
unbreak redirect

Commit:
copy edits

alexfornuto commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
````diff
@@ -0,0 +1,452 @@
+your kubeconfig:
+
+```code
+$ tsh kube login minikube
````
As above, should this be `kubectl`?

ptgott replied:
I've verified that this command is correct

Commit:
more copy edits

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -2634,7 +2634,8 @@ tctl version
 ## Resource filtering
 
-Both `tsh` and `tctl` allow you to filter `nodes`, `apps`, `db`, and `kube` resources using the `--search` and `--query` flags.
+Both `tsh` and `tctl` allow you to filter `nodes`, `apps`, `db`, and `kube`
```
Should we add `windows_desktop` to this list?

ptgott replied:
Added. I've also used regular English instead of code style for these terms since `apps`, `nodes`, etc. aren't actually mentioned in the `--search` flag.

alexfornuto reviewed (commented):
Added some copy edits directly, left more suggestions below.

ptgott commented on docs/pages/access-controls/access-requests/resource-requests.mdx:
```diff
@@ -323,52 +340,182 @@ spec:
+
+#### `pod`
+
+You can restrict access to creating `pod` resources by assigning values to the
+`kubernetes_resources` field in the `spec.allow` or `spec.deny` fields. 
+
+The following role allows access to Kubernetes pods with the name `nginx` in any
+namespace, and all pods in the `dev` namespace:
```
Oops, that "creating" shouldn't be there. Changed

ptgott commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
````diff
@@ -0,0 +1,452 @@
+Ensure that the resources are deployed:
+
+```code
+$ kubectl -n development get pods
+$ kubectl -n production get pods
````
The text below this describes the expected output. If it makes sense, I can fire up the demo again and copy/paste the return value

alexfornuto replied:
meh, not critical, just something to keep in mind in the future.

ptgott commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
````diff
@@ -0,0 +1,452 @@
+Verify that the `teleport` pod is running in your cluster:
+
+```code
+$ kubectl -n teleport get pods
````
I think for now, the "Verify that the `teleport` pod..." instructions are enough here. The user will see the `teleport` pod in the output

ptgott commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
````diff
@@ -0,0 +1,452 @@
+following command:
+
+```code
+$ tctl create kube-access.yaml
````
Since this is the first time we create the role in this page, I didn't include the `-f`/`--force` flag. Maybe we should leave it out in case the user has accidentally defined a `kube-access` role resource already?

alexfornuto replied:
Fair enough. I only suggested it because I thought we were using `-f` as a convention, but your argument against it in this case is valid.

ptgott commented on docs/pages/kubernetes-access/manage-access/rbac.mdx:
````diff
@@ -0,0 +1,452 @@
+List the Kubernetes clusters you can access via Teleport:
+
+```code
+$ tsh kube ls
````
`tsh kube ls` is the correct command

ptgott replied:
I've verified that this command is correct

ptgott committed:
Respond to alexfornuto feedback
